control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs

```diff
@@ -0,0 +1,146 @@
+        let mut results = vec![];
+        let nodes = context.specs().get_nodes();
+
+        for node in nodes {
+            results.push(check_and_drain_node(context, &node).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+
+    async fn poll_event(&mut self, context: &PollContext) -> bool {
+        matches!(context.event(), PollEvent::TimedRun)
+    }
+}
+
+/// Republish the specified volume on any node other than its current node
+async fn republish_volume(context: &PollContext, vol_uuid: &VolumeId) -> Result<(), SvcError> {
+    let request = RepublishVolume::new(vol_uuid.clone(), None, VolumeShareProtocol::Nvmf, false);
+    tracing::info!(
+        volume.uuid = vol_uuid.as_str(),
+        "Attempting to republish volume"
+    );
+
+    let mut volume = context.registry().specs().volume(&request.uuid).await?;
+    volume.republish(context.registry(), &request).await?;
+
+    Ok(())
+}
+
+/// Returns true if the specified node is in the draining state
+fn is_node_draining(context: &PollContext, nodeid: &NodeId) -> Result<bool, SvcError> {
+    let nodespec = match context.specs().get_node(nodeid) {
+        Ok(ns) => ns,
+        Err(e) => {
+            return Err(e);
+        }
+    };
+    Ok(nodespec.is_draining())
+}
+
+/// Drain the specified node if in draining state
+async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> PollResult {
+    if node_spec.is_draining() {
+        let node_id = node_spec.id().clone();
+        tracing::trace!(node.id = node_spec.id().as_str(), "Draining node");
+        // iterate through all the volumes, find those with a nexus hosted on the node
+        let volumes = context.registry().get_volumes().await;
+
+        for volume in volumes.iter() {
+            if volume.spec().target.is_some()
+                && volume.spec().target.unwrap().node().clone() == node_id
+            {
+                let nexus_id = volume.spec().target.unwrap().nexus().clone();
+                let vol_id = volume.spec().uuid;
+                tracing::info!(
+                    volume.uuid = vol_id.as_str(),
+                    nexus.uuid = nexus_id.as_str(),
+                    node.id = node_spec.id().as_str(),
+                    "Moving volume"
+                );
+                match republish_volume(context, &vol_id).await {
+                    Ok(()) => {}
+                    Err(e) => {
+                        tracing::error!(
+                            "Failed to republish volume {} on node {}, error {}",
+                            vol_id,
+                            node_id,
+                            e,
+                        );
+                        return PollResult::Err(e);
+                    }
+                };
+                // wait for the nexus to be removed, or for the drain to be cancelled
```

Comment:
the reconcilers don't quite work like this atm, they don't wait inline, rather they try again next time they run...
But I don't think we need it at all here because the republish already succeeds when it completes, so it's already moved.

Reply:
From the logs, when testing this, republish_volume() seems to return before the the nexus is moved. The loop (line 105) runs 3 or 4 times before the nexus is found to be gone, so it seems to be asynchronous AFAICT.
If I do remove the loop, I suppose there is a risk that republish_volume will be called multiple times for the same volume (i.e. when the reconciler gets triggered again) if it has not yet been moved from the node by that time. Is that safe?

Reply:
Removed wait loop as superfluous.